### ISSUE TITLE
1.14.5 Native patching behavior change - XR desired state dropped if a resource won't get all required schema values 

### ISSUE THREAD

**Issue by y4ri** [crossplane/crossplane#5292]
### What happened?

When one of the resources in composition has a patch to a field that in its schema is required, but the patch lacks `fromFieldPath: Required` policy, XR desired state from other resources patches will be dropped which can lead to deadlock and stopping composition progress. 

This behavior changed, in 1.13.x the resource that lacks required field fails but the desired state of other patches to the XR would be saved and the resource would be created on next reconciliation.

### How can we reproduce it?

Apply manifests below, xrd->compositions->claim

On 1.14.5 it will create second XDummy and in summary 2 DeploymentRuntimeConfig only after adding `fromFieldPath: Required` patch for required value .

Example demo xrd, compositions and claim file:

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata: 
  name: xdummys.example.org
spec:
  defaultCompositionRef:
    name: test-comp-two
  group: example.org
  names:
    kind: XDummy
    plural: xdummys
  claimNames:
    kind: Dummy
    plural: dummys
  versions:
  - name: v1alpha1
    served: true
    referenceable: true
    schema:
      openAPIV3Schema:
        type: object
        properties:
          spec:
            type: object
            properties:
              exampleField:
                type: string  
            required:
            - exampleField
```

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: test-comp-one
spec:
  compositeTypeRef:
    apiVersion: example.org/v1alpha1
    kind: XDummy
  resources:
    - name: druntime-one
      base:
        apiVersion: pkg.crossplane.io/v1beta1
        kind: DeploymentRuntimeConfig
        metadata:
          annotations:
            crossplane.io/external-name: druntime-one
        spec: {}
      patches:
      # desired state of this patch will be saved as it is patching MR
      - type: FromCompositeFieldPath
        fromFieldPath: spec.exampleField
        toFieldPath: metadata.annotations.stringFromClaim
      # this one will fail to be saved
      - type: ToCompositeFieldPath
        fromFieldPath: metadata.name
        toFieldPath: metadata.annotations.exampleVal

    - name: second-xdummy
      base:
        apiVersion: example.org/v1alpha1
        kind: XDummy
        metadata:
          annotations:
            crossplane.io/external-name: dummy-two
        spec: {}
      patches:
      - type: FromCompositeFieldPath
        fromFieldPath: metadata.annotations.exampleVal
        toFieldPath: spec.exampleField
        # try without this policy first, 1.13.2 succeeds, 1.14.5 deadlocks
        # policy:
        #   fromFieldPath: Required
```

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: test-comp-two
spec:
  compositeTypeRef:
    apiVersion: example.org/v1alpha1
    kind: XDummy
  resources:
    - name: druntime-two
      base:
        apiVersion: pkg.crossplane.io/v1beta1
        kind: DeploymentRuntimeConfig
        metadata:
          annotations:
            crossplane.io/external-name: druntime-two
        spec: {}
      patches:
      - type: ToCompositeFieldPath
        fromFieldPath: metadata.name
        toFieldPath: metadata.annotations.exampleVal
```

```yaml
apiVersion: example.org/v1alpha1
kind: Dummy
metadata:
  name: test
spec:
  exampleField: "test string"
  compositionRef:
    name: test-comp-one
```

### What environment did it happen in?
Crossplane version: 1.14.5

Could replicate on eks 1.24 as well as 1.27 kind cluster.

**Comment by negz** [comment]
@y4ri Do you see any errors or warnings when this is happening? If so, could you let us know what they are.

**Comment by y4ri** [comment]
1.14.5 - On XR level just a bunch of

```bash
Warning  ComposeResources         13s                defined/compositeresourcedefinition.apiextensions.crossplane.io  cannot compose resources: cannot apply composed resource: cannot create object: XDummy.example.org "test-qksw8-dlpcx" is invalid: spec.exampleField: Required value
```

ending with

```bash
Warning  ComposeResources         9s (x14 over 13s)  defined/compositeresourcedefinition.apiextensions.crossplane.io  (combined from similar events): cannot compose resources: cannot apply composed resource: cannot create object: XDummy.example.org "test-qksw8-m96gr" is invalid: spec.exampleField: Required value
```

On 1.13.2 I only get 1 warrning:

```bash
Warning  ComposeResources         21s                defined/compositeresourcedefinition.apiextensions.crossplane.io  composed resource "second-xdummy": cannot use dry-run create to name composed resource: XDummy.example.org "test-jf7q7-5nswr" is invalid: spec.exampleField: Required value
```

**Comment by xoanmi** [comment]
We have the same problem. We have not been able to work around it. Did you do it @y4ri ?

**Comment by y4ri** [comment]
@xoanmi just like in "test-comp-one" composition above, you need to add

```yaml
policy:
  fromFieldPath: Required
 ```
 
 to any patches of XRs/MRs in composition that patch in the value required by schema.
 
 With an example of VPC Managed Resource from [aws provider](https://marketplace.upbound.io/providers/crossplane-contrib/provider-aws/v0.46.0/resources/ec2.aws.crossplane.io/VPC/v1beta1#doc:spec-forProvider-cidrBlock) you would need that policy for a patch responsible for value of `spec.forProvider.cidrBlock`

**Comment by xoanmi** [comment]
> @xoanmi just like in "test-comp-one" composition above, you need to add
> 
> ```yaml
> policy:
>   fromFieldPath: Required
> ```
> 
> to any patches of XRs/MRs in composition that patch in the value required by schema.
> 
> With an example of VPC Managed Resource from [aws provider](https://marketplace.upbound.io/providers/crossplane-contrib/provider-aws/v0.46.0/resources/ec2.aws.crossplane.io/VPC/v1beta1#doc:spec-forProvider-cidrBlock) you would need that policy for a patch responsible for value of `spec.forProvider.cidrBlock`

Thanks! I misread your previous message. We have applied the workaround and it works! 🚀

**Comment by phisco** [comment]
I think this is a duplicate of https://github.com/crossplane/crossplane/issues/5114

**Comment by negz** [comment]
> I think this is a duplicate of https://github.com/crossplane/crossplane/issues/5114

@phisco How sure are you? At a glance it sounds like the described behaviour is different.
